src/Orleans/Utils/IKeyedServiceCollection.cs

```diff
@@ -10,7 +12,40 @@ namespace Orleans.Runtime
     public interface IKeyedServiceCollection<in TKey, out TService>
         where TService : class
     {
-        TService GetService(TKey key);
+        TService GetService(IServiceProvider services, TKey key);
+    }
+
+    public interface IKeyedService<TKey, out TService> : IEquatable<TKey>
+    {
+        TService GetService(IServiceProvider services);
+    }
+
+    public class KeyedService<TKey, TService, TInstance> : IKeyedService<TKey, TService>
+            where TInstance : TService
+    {
+        private readonly TKey key;
+
+        public KeyedService(TKey key)
+        {
+            this.key = key;
+        }
+
+        public TService GetService(IServiceProvider services) => services.GetService<TInstance>();
```

Comment:
The more I look at this approach (which is similar to what we talked about before we protoyped [Orleans hosting](https://github.com/jdom/Orleans.Hosting/blob/master/Orleans.Hosting/)), the less convinced I am that we should be putting the name mapping itself as singleton services in DI.
I think the previously proposed approach of separating the building of the individual services from the keyed service collection itself is useful. After all, there is no need to have the factory be a transient service, as the grain activation context could (and IMO should, regardless of this proposed change) be passed in when creating the concrete storage feature, and not be injected in the constructor.
If we make that minor change (having the service not be transient), then it opens up a much cleaner separation. An example of that potential separation can be seen here (notice that the key->service mapping is not part of the more general purpose ServiceCollection registrations): https://github.com/jdom/Orleans.Hosting/blob/master/Orleans.Hosting/KeyedServiceCollection.cs

If you want, I can submit a small PR adapting that old branch to your branch specifically, as there is some small logic that is not yet available in our current silo initialization, which is the step to configure the silo after the IServiceProvider has been built.

Otherwise, I'm fine to not do this full change now and potentially improve it later, once we add that mentioned capability (which we'll have at least when we incorporate the SiloBuilder). Nevertheless it would be nice to at least not have the `INamedExampleStorageFactory` implementations be transient services, but instead have the `INamedExampleStorageFactory` receive the `IGrainActivationContext` on every call to `Create`. I understand this factory is in test code, but the support for transient keyed services is not, as well as this is defining how the real implementations will use this feature.